lsp/roslyn_ls.lua

```diff
@@ -0,0 +1,195 @@
+--- borrowed some source from: https://github.com/seblj/roslyn.nvim
+-- config that activates keymaps and enables snippet support
+---comment
+
+local uv = vim.uv
+local fs = vim.fs
+
+---@param client vim.lsp.Client
+---@param target string
+local function on_init_sln(client, target)
+  vim.notify("Initializing: " .. target, vim.log.levels.INFO, { title = "roslyn-ls" })
+  ---@diagnostic disable-next-line: param-type-mismatch
+  client:request("solution/open", {
+      solution = vim.uri_from_fname(target),
+  },
+  function (err, result, context)
+  end)
+end
+
+---@param client vim.lsp.Client
+---@param project_files string[]
+local function on_init_project(client, project_files)
+  vim.notify("Initializing: projects", vim.log.levels.INFO, { title = "roslyn-ls" })
+  ---@diagnostic disable-next-line: param-type-mismatch
+  client:request("project/open", {
+      projects = vim.tbl_map(function(file)
+          return vim.uri_from_fname(file)
+      end, project_files),
+  },
+  function (err, result, context)
+  end)
+end
+
+local function roslyn_handlers()
+  return {
+        ["workspace/projectInitializationComplete"] = function(_, _, ctx)
+            vim.notify("Roslyn project initialization complete", vim.log.levels.INFO, { title = "roslyn.nvim" })
+
+            local buffers = vim.lsp.get_buffers_by_client_id(ctx.client_id)
+            for _, buf in ipairs(buffers) do
+                vim.lsp.util._refresh("textDocument/diagnostic", { bufnr = buf })
+            end
+
+            ---NOTE: This is used by rzls.nvim for init
+            vim.api.nvim_exec_autocmds("User", { pattern = "RoslynInitialized", modeline = false })
+        end,
+        ["workspace/_roslyn_projectHasUnresolvedDependencies"] = function()
+            vim.notify("Detected missing dependencies. Run dotnet restore command.", vim.log.levels.ERROR, {
+                title = "roslyn.nvim",
+            })
+            return vim.NIL
+        end,
+        ["workspace/_roslyn_projectNeedsRestore"] = function(_, result, ctx)
+            local client = assert(vim.lsp.get_client_by_id(ctx.client_id))
+
+            ---@diagnostic disable-next-line: param-type-mismatch
+            client:request("workspace/_roslyn_restore", result, function(err, response)
+                if err then
+                    vim.notify(err.message, vim.log.levels.ERROR, { title = "roslyn.nvim" })
```

Comment:
The title for `vim.notify` calls should at least be the same for all calls if we should have it